shared/AppInsightsCore/src/JavaScriptSDK/BaseCore.ts

```diff
@@ -249,14 +265,182 @@ export class BaseCore implements IAppInsightsCore {
+                _self.config = null;
+                _self._extensions = [];
+
+                _telemetryInitializerPlugin = new TelemetryInitializerPlugin();
+                _eventQueue = [];
+                _notificationManager = null;
+                _perfManager = null;
+                _cfgPerfManager = null;
+                _cookieManager = null;
+                _pluginChain = null;
+                _coreExtensions = null;
+                _configExtensions = null;
+                _channelControl = null;
+                _channelConfig = null;
+                _channelQueue = null;
+                _internalLogsEventName = null;
+            }
+
+            // Initialize or Re-initialize the plugins
+            function _initPluginChain(config: IConfiguration) {
+                // Extension validation
+                let theExtensions = _validateExtensions(_self.logger, ChannelControllerPriority, _configExtensions);
+            
+                _coreExtensions = theExtensions.core;
+                _pluginChain = null;
+            
+                // Sort the complete set of extensions by priority
+                let allExtensions = theExtensions.all;
+
+                // Initialize the Channel Queues and the channel plugins first
+                _channelQueue = objFreeze(createChannelQueues(_channelConfig, allExtensions, config, _self));
+                _channelControl = createChannelControllerPlugin(_channelQueue, _self);
+
+                // Add on "channelController" as the last "plugin"
+                allExtensions.push(_channelControl);
+                _coreExtensions.push(_channelControl);
+
+                // Required to allow plugins to call core.getPlugin() during their own initialization
+                _self._extensions = sortPlugins(allExtensions);
+
+                // Initialize the controls
+                _channelControl.initialize(config, _self, allExtensions);
+                
+                initializePlugins(_self.getProcessTelContext(), allExtensions);
+
+                // Now reset the extensions to just those being managed by Basecore
+                _self._extensions = objFreeze(sortPlugins(_coreExtensions || [])).slice();
+            }
+
+            function _getPlugin<T extends IPlugin = IPlugin>(pluginIdentifier: string): ILoadedPlugin<T> {
+                let theExt: ILoadedPlugin<T> = null;
+                let thePlugin: IPlugin = null;
+
+                arrForEach(_self._extensions, (ext: any) => {
+                    if (ext.identifier === pluginIdentifier) {
+                        thePlugin = ext;
+                        return -1;
+                    }
+                });
+
+                if (thePlugin) {
+                    theExt = {
+                        plugin: thePlugin as T
+                    }
+                }
+
+                return theExt;
+            }
+
+            function _getPluginChain() {
+                if (!_pluginChain) {
```

Comment:
Just a helper, to lazily create the current chain to avoid initializing a new one every time something changes even if now events are processed.

- Think adding or removing multiple plugins on the fly.